File: ts-plugins/src/idempotency/idempotent-rich-resume-executor.ts
import type { FlowEngine, FlowDefinition } from '@unlaxer/tramli';
import type { IdempotencyRegistry, CommandEnvelope } from './types.js';
import { RichResumeExecutor, type RichResumeResult } from '../resume/rich-resume.js';

export class IdempotentRichResumeExecutor {
  private readonly registry: IdempotencyRegistry;
  private readonly delegate: RichResumeExecutor;

  constructor(engine: FlowEngine, registry: IdempotencyRegistry) {
    this.registry = registry;
    this.delegate = new RichResumeExecutor(engine);
  }

  async resume<S extends string>(
    flowId: string,
    definition: FlowDefinition<S>,
    envelope: CommandEnvelope,
    knownBeforeState: S,
  ): Promise<RichResumeResult<S>> {
    if (!this.registry.markIfFirstSeen(flowId, envelope.commandId)) {
      return {
        status: 'ALREADY_COMPLETE',
        error: new Error(`duplicate commandId ${envelope.commandId}`),
      };
    }
    return this.delegate.resume(flowId, definition, envelope.externalData, knownBeforeState);
  }
}

File: java-core-plugins/src/main/java/org/unlaxer/tramli/plugins/idempotency/IdempotentRichResumeExecutor.java
package org.unlaxer.tramli.plugins.idempotency;

import org.unlaxer.tramli.FlowDefinition;
import org.unlaxer.tramli.FlowEngine;
import org.unlaxer.tramli.FlowState;
import org.unlaxer.tramli.plugins.resume.RichResumeExecutor;
import org.unlaxer.tramli.plugins.resume.RichResumeResult;
import org.unlaxer.tramli.plugins.resume.RichResumeStatus;

public final class IdempotentRichResumeExecutor {
    private final IdempotencyRegistry registry;
    private final RichResumeExecutor delegate;

    public IdempotentRichResumeExecutor(FlowEngine engine, IdempotencyRegistry registry) {
        this.registry = registry;
        this.delegate = new RichResumeExecutor(engine);
    }

    public <S extends Enum<S> & FlowState> RichResumeResult<S> resume(String flowId,
                                                                       FlowDefinition<S> definition,
                                                                       CommandEnvelope envelope,
                                                                       S knownBeforeState) {
        if (!registry.markIfFirstSeen(flowId, envelope.commandId())) {
            return new RichResumeResult<>(RichResumeStatus.ALREADY_COMPLETED, knownBeforeState, knownBeforeState,
                    "duplicate commandId " + envelope.commandId(), null);
        }
        return delegate.resume(flowId, definition, envelope.externalData(), knownBeforeState);
    }
}
